Introduce ArgumentBuilder

drewnoakes opened the pull request:
Fixes #4661.

Changes proposed in this pull request:
 - Add new `ArgumentBuilder` class, with API docs and tests
 - Use new class in places where argument construction logic was non-trivial
 
What did I do to test the code and ensure quality:
 - Added unit tests for `ArgumentBuilder`
 - Unit tests
 - Manual tests

The unified diff makes 33f483c a bit hard to read. Easiest to review in the side-by-side view if you have a wide enough screen.

Has been tested on (remove any that don't apply):
 - GIT 2.16
 - Windows 10

gerhardol commented on GitCommands/Git/GitModule.cs:
```diff
@@ -3360,14 +3280,19 @@ public IEnumerable<string> GetPreviousCommitMessages(string revision, int count)
+            RunGitCmdDetached(args.ToString());
 
-            return output;
+            // TODO why bother returning this?
```
This is supposed to return any possible error message from the difftool
Maybe we can remove the return

drewnoakes replied:
I see. `RunGitCmdDetached` doesn't wait for any process output anyway, so yes it seems redundant. Open questions exist about whether something should track the returned `Process` object, but that feels like a different PR.

I am doing a bunch of work towards the goal of having a single point through which all git executions occur. I can cover this under that chunk of work.

For now I've extended the TODO comment.

drewnoakes commented on GitCommands/Git/GitModule.cs:
```diff
@@ -1661,40 +1649,34 @@ public string GetPuttyKeyFileForRemote(string remote)
-                pullArgs = "--progress ";
-            }
-
-            if (rebase)
```
Well spotted. I missed the reassignment. Fixed.

Aside: this is an example of the kind of behaviour that'd be easier to notice if we had [readonly locals](https://github.com/dotnet/csharplang/issues/188) in C#.

drewnoakes replied:
Turns out I didn't fix it property as the tests now fail. Will push up another fix shortly.

drewnoakes replied:
Reverted the change here and the test passes again.

gerhardol replied:
The changed behavior should be the correct.

drewnoakes commented on GitCommands/Git/GitModule.cs:
```diff
@@ -2216,45 +2163,19 @@ public IList<PatchFile> GetRebasePatchFiles()
+                { noVerify, "--no-verify" },
+                { signOff, "--signoff" },
+                { !string.IsNullOrEmpty(author), $"--author=\"{author?.Trim().Trim('"')}\"" },
+                { gpgSign && string.IsNullOrWhiteSpace(gpgKeyId), "-S" },
```
I believe this change to be correct. Let me know if I'm missing something.

Before:

```c#
if (gpgSign)
{
    command += " -S";
    if (!string.IsNullOrWhiteSpace(gpgKeyId))
    {
        command += gpgKeyId;
    }
}
```

After:

```c#
{ gpgSign && string.IsNullOrWhiteSpace(gpgKeyId), "-S" },
{ gpgSign && !string.IsNullOrWhiteSpace(gpgKeyId), $"-S{gpgKeyId}" },
```

Both of those expressions cannot be true. Note that they both include `-S`.

In English, if `gpgSign` is true, then depending upon `string.IsNullOrWhiteSpace(gpgKeyId)` either `-S` or `-S<keyid>` is added.

gerhardol replied:
Correct. In this situation it is more confusing due to space automatically added between the arguments.

drewnoakes commented:
Fixed build, rebased and added usage of `ArgumentBuilder` in a few more places in `GitCommandHelpers`. They should be quick to review (last three commits). Can break them off into a different PR if preferred.

I don't plan on adding anything more here beyond review feedback.

gerhardol commented on GitCommands/Git/GitCommandHelpers.cs:
```diff
@@ -465,12 +465,13 @@ private static string SubmoduleUpdateCommand(string name)
-                return "submodule sync";
-            }
+                "submodule sync",
+                name?.Trim()
```
QuoteNE()

gerhardol commented on GitCommands/Git/GitCommandHelpers.cs:
```diff
@@ -544,60 +539,38 @@ public static string CloneCmd(string fromPath, string toPath)
+                { isSingleBranch == false, "--no-single-branch" },
+                "--progress",
+                { branch == null, "--no-checkout" },
+                { branch != "", $"--branch {branch}" },
```
branch==null is added here too?

gerhardol commented on GitCommands/Git/GitCommandHelpers.cs:
```diff
@@ -690,73 +654,37 @@ public static string PushMultipleCmd(string remote, IEnumerable<GitPushAction> p
+                { VersionInUse.PushCanAskForProgress, "--progress" },
+                path.ToPosixPath().Trim().Quote(),
+                { all, "--tags" },
+                { !all, $"tag {tag.Replace(" ", "")}" }
```
if (!string.IsNullOrEmpty(tag))

gerhardol commented on GitCommands/Git/GitCommandHelpers.cs:
```diff
@@ -690,73 +654,37 @@ public static string PushMultipleCmd(string remote, IEnumerable<GitPushAction> p
+            };
 
-            return sforce;
+            return args.ToString();
```
Not empty if (!string.IsNullOrEmpty(tag)) or !all

drewnoakes replied:
I'm going to add this to the beginning of the method:

```c#
if (!all && string.IsNullOrWhiteSpace(tag))
{
    // TODO this is probably an error
    return "";
}
```

gerhardol commented on GitCommands/Git/GitCommandHelpers.cs:
```diff
@@ -776,94 +704,48 @@ public static string StartBisectCmd()
-            }
+                "bisect",
+                bisectOption,
+                revisions
```
revisions.Length==0

drewnoakes replied:
In this case it's fine because the relevant `Add` method doesn't modify the argument string for null or empty enumerables (nor for null or white-space items they produce).

gerhardol replied:
correct. My head got spinning just by reviewing this

gerhardol commented on GitCommands/Git/GitCommandHelpers.cs:
```diff
@@ -776,94 +704,48 @@ public static string StartBisectCmd()
+                { interactive && !autosquash, "--no-autosquash" },
+                { preserveMerges, "--preserve-merges" },
+                { autostash, "--autostash" },
+                from.QuoteNE(),
```
there is no check if just one of from and onto is null. Should at least be a comment

drewnoakes replied:
I've added a precondition:

```c#
if (from == null ^ onto == null)
{
    throw new ArgumentException(
        $"For arguments \"{nameof(from)}\" and \"{nameof(onto)}\", either both must have values, or neither may.");
}
```

Existing usages will be fine with this.

drewnoakes commented:
@gerhardol thanks for the awesome review. I believe your points have been addressed now.

gerhardol commented:
Great rewrite. I will let @RussKie comment too.

RussKie commented on GitCommands/ArgumentBuilder.cs:
```diff
@@ -0,0 +1,260 @@
+            else if (mode == IgnoreSubmodulesMode.All)
+            {
+                builder.Add("--ignore-submodules=all");
+            }
```
I am not particularly fond of so many `if/else` as I can't say I find it easier to read than a `switch` statement.
Besides `switch` typically has an edge over multiple `if/else` - https://stackoverflow.com/a/767840.

Beside in a `switch` it is easier to make default handling more obvious. And the default case can be leveraged for testing too:

```
switch (mode)
{
    // ...
    default: throw new NotSupportedException("developer you suck!");
}
```

And have a unit test to ensure that whenever the enum definition changes no exception is thrown. 
For example (pseudocode):

```
[Test]
public void Test_all_enum_values_mapped()
{
    var values = Enum.GetValues<EnumType>();
    foreach (var value in values)
    {
        ArgumentBuilderExtensions.Add(new ArgumentBuilder(), (EnumType)value).Should().NotThrow();
    }
}
```

RussKie commented on GitUI/CommandsDialogs/FormRebase.cs:
```diff
@@ -233,15 +233,15 @@ private void OkClick(object sender, EventArgs e)
-                                                        chkStash.Checked);
+                rebaseCmd = GitCommandHelpers.RebaseCmd(
+                    Branches.Text, chkInteractive.Checked, chkPreserveMerges.Checked,
+                    chkAutosquash.Checked, chkStash.Checked);
```
Arguably the readability decreased...

RussKie commented on GitCommands/ArgumentBuilder.cs:
```diff
@@ -0,0 +1,260 @@
+        }
+    }
+
+    public static class ArgumentBuilderExtensions
```
Please move it to a separate file, it makes easier to locate files outside IDE.
Please also have associated tests moved to a matching file.

drewnoakes replied:
Done.

drewnoakes commented on GitCommands/ArgumentBuilder.cs:
```diff
@@ -0,0 +1,260 @@
+        /// <param name="values">A sequence of strings to add.</param>
+        public static void Add(this ArgumentBuilder builder, [CanBeNull, ItemCanBeNull] IEnumerable<string> values)
+        {
+            if (values != null)
```
Done.

drewnoakes commented on GitCommands/ArgumentBuilder.cs:
```diff
@@ -0,0 +1,260 @@
+        /// <param name="ifConditionTrue">A sequence of strings to add if <paramref name="condition"/> is <c>true</c>.</param>
+        public static void Add(this ArgumentBuilder builder, bool condition, [CanBeNull, ItemCanBeNull] IEnumerable<string> ifConditionTrue)
+        {
+            if (condition && ifConditionTrue != null)
```
Ditto.

gerhardol commented:
@RussKie 
I support merging this also before  adding more tests

RussKie commented:
Duly noted. How do we know that we haven't introduced a regression?

I'd hate to see the next release marred with "_git command isn't doing what I've asked. It used to work in 2.51_" type of issues...

gerhardol commented:
> How do we know that we haven't introduced a regression?

We cannot be sure. But keeping the speed up will get us in a better position for future releases.
Whatever keeps @drewnoakes working....

drewnoakes commented:
I prepended some commits that tested the implementation of command generation before switching to `ArgumentBuilder`. They flagged one problem with the construction of push args when specifying branches to push from and to. So we prevented a regression, and got some tests to boot.

RussKie commented:
awesome, I'll have a look tonight

On Thu, 22 Mar 2018 7:04 am Gerhard Olsson <notifications@github.com> wrote:

> *@gerhardol* approved this pull request.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/gitextensions/gitextensions/pull/4662#pullrequestreview-105886091>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AEMyXgg4p_qzfXk5ltIwcKwJjjTENqbZks5tgrJUgaJpZM4SvN80>
> .
>

drewnoakes commented:
Rebased to fix conflicts.

drewnoakes commented:
Rebased to fix conflicts.

Would be great to get this in so I can put together the PR around improved git process execution and management.

sharwell commented on GitCommands/ArgumentBuilderExtensions.cs:
```diff
@@ -0,0 +1,204 @@
+    {
+        /// <summary>
+        /// Adds <paramref name="ifConditionTrue"/> to the argument list if <paramref name="condition"/>
+        /// is <c>true</c>.
```
:bulb: Most of the places where `true`, `false`, and `null` are used in documentation could benefit from the `<see langword="true"/>` syntax. This is supported directly by recent versions of Visual Studio, so if you just type `true` and press <kbd>Ctrl</kbd>+<kbd>Space</kbd> it will expand to the correct form.

drewnoakes replied:
I've seen this in the past but wondered about the benefit. It's semantically clearer to a computer, but not necessarily to a human. Are there tools that benefit from this?

sharwell replied:
1. It's much easier to type (code completion)
2. QuickInfo in Visual Studio uses it for highlighting
3. Sandcastle Help File Builder uses it if you produce documentation output
4. It would be the basis for keyword handling in other tools as well

drewnoakes replied:
I agree if you're comparing with a plain keyword, but `<c>null</c>` is going to be formatted the same, and enjoys the same completion, at least with my setup :)

RussKie replied:
`<c>null</c>` will be rendered as monospace, but it won't be converted for other languages the same way `<see langword="null" />` will - `null` is a c# keyword, for vs it will get rendered as `Nothing`

sharwell commented on GitCommands/ArgumentBuilder.cs:
```diff
@@ -0,0 +1,82 @@
+    /// };
+    /// </code>
+    /// </example>
+    public sealed class ArgumentBuilder : IEnumerable
```
💭 This could probably be a `struct` to avoid the extra allocations.

drewnoakes replied:
Good idea. Done.

drewnoakes replied:
I had to back this out as it broke logic. The extension methods accept `ArgumentBuilder` as a parameter, which ends up being a copy. The `this` param cannot be `ref` unfortunately. That'd be useful for structs.

This is marked as intended for 7.2, but that ship has sailed. Am I missing something?

https://github.com/dotnet/csharplang/issues/186

sharwell replied:
What error were you seeing? For this type, even copies should behave the same as the original.

sharwell commented on GitCommands/Git/GitCommandHelpers.cs:
```diff
@@ -378,45 +379,40 @@ private static string SubmoduleUpdateCommand(string name)
             {
-                return "submodule sync";
-            }
+                "submodule sync",
```
❗️ I would expect an argument build to automatically quote arguments with spaces when necessary, which means I would expect to see this as two arguments instead of one.

sharwell commented on GitCommands/Git/GitCommandHelpers.cs:
```diff
@@ -455,63 +446,41 @@ public static string CloneCmd(string fromPath, string toPath)
-            var command = lfs ? "lfs clone " : "clone ";
+            var args = new ArgumentBuilder
+            {
+                { lfs, "lfs" },
```
💭 This is short, but I'm 100% convinced it's better

sharwell replied:
Well I see I totally messed up my comment by forgetting 'not'

RussKie replied:
I wondered whether it was the case... :)

sharwell commented on GitCommands/ArgumentBuilderExtensions.cs:
```diff
@@ -0,0 +1,204 @@
+        /// <param name="builder">The <see cref="ArgumentBuilder"/> to add arguments to.</param>
+        /// <param name="condition">Whether or not to add <paramref name="ifConditionTrue"/> to the argument list.</param>
+        /// <param name="ifConditionTrue">The string to add if <paramref name="condition"/> is <c>true</c>.</param>
+        public static void Add(this ArgumentBuilder builder, bool condition, [CanBeNull] string ifConditionTrue)
```
:bulb: Consider making the multiple-argument forms take a tuple type instead of simply multiple arguments. This will allow use of named arguments at the call site with compiler validation.

```
public static void Add(this ArgumentBuilder builder, (bool condition, string ifConditionTrue) value)
```

```
new ArgumentBuilder
{
  (condition: true, ifConditionFalse: "") // Compiler error
}
```

RussKie replied:
Could you please explain how the above causes a compiler warning

drewnoakes replied:
Because of the label mismatch. `ifConditionTrue` != `ifConditionFalse`.

sharwell commented on GitCommands/ArgumentBuilder.cs:
```diff
@@ -0,0 +1,82 @@
+    /// </summary>
+    /// <remarks>
+    /// To retrieve the constructed argument list string, call <see cref="ToString"/>.
+    /// <para />
```
:memo: `<para>` should wrap the text, and the line between paragraphs should be empty, like this:

```
/// <remarks>
/// <para>To retrieve the constructed argument list string, call <see cref="ToString"/>.</para>
///
/// <para>Arguments are separated by a single space character.</para>
///
/// <para>Adding <see langword="null"/> or white-space strings has no effect on the output, which can be
/// useful in some calling constructions.</para>
///
/// ...
/// </remarks>
```

drewnoakes replied:
Again this is more semantically correct but I do find it harder to read, and it does seem to be less common in the wild, at least on codebases I've worked on. It does seem more common in GE though. Will change.

drewnoakes replied:
Changed and pushed.

drewnoakes commented:
> Requesting changes that arguments with spaces are handled automatically by the builder, if/when they appear

Interesting idea. I'll have to review this tomorrow as it'll mean looking at all current usages. But it should end up being clearer and less prone to errors.

`StringExtensions.Quote` does not escape inner quotes. I'm not sure it can know how to escape either. In an argument builder you do know the context, so it does feel like the right place to be quoting and escaping.

Other than escaping quotes in strings, are there any other cases to consider?

RussKie commented:
Good work

drewnoakes commented:
> Requesting changes that arguments with spaces are handled automatically by the builder, if/when they appear

I started looking at this and came up with a concern. I wonder about a form like this:

```text
git foo --bar="two words"
```

If that form exists, how do we handle it, given currently the `--bar="two words"` string must be a single string argument to `ArgumentBuilder`? The quoting logic would likely convert it into something like `"--bar=\"two words\""`. That might work, but I'm not sure and it requires further testing. It's possible that this form doesn't exist, but I'm not confident. Anyone know?

If so, we could come up with some rules about strings that start with `--` and have an `=`. Or add a new overload for a=b arguments. Maybe. At some point the `ArgumentBuilder` could become a bit too magical or complex. One downside of the implicit `Add` calls is that there's no point to inspect documentation in the IDE for the resolved overload. You just have to know how it works by reading the docs on the type itself and do the overload resolution in your head which can get hard to do.

I'm conscious of time zones. It's morning here, and @RussKie's likely off to bed soon. I'd like to look at getting that git process management PR together today and this PR changes a lot of the same code. I'm happy to look at automatic quoting of spaces in arguments in a subsequent PR. If this is otherwise good to go then merging it now would make me more productive today.

RussKie commented:
...and it is probably late night where Sam is. :)

I'm quite happy with the current proposal. Whilst Sam's comments certainly have value, they may complicate the implementation. Let's note these comments, maybe as TODOs, and keep moving.

sharwell commented:
The biggest issue I had is the inability to use named arguments. Switching to tuples would allow this.

RussKie commented:
@drewnoakes is that easy to do?

drewnoakes commented:
I will look into it. I can't see how tuples would help as the names of tuple arguments don't seem to affect overload resolution. In fact, the following doesn't compile:

```c#
public void Foo((string s, int i) tuple)
{

}

public void Foo((string s, int j) tuple) // <-- error here
{

}
```

> error CS0111: Type 'Bar' already defines a member called 'Foo' with the same parameter types

@sharwell let me know if I'm missing something.

Either way, if you're happy with this PR and can merge before turning it, I will be able to do more throughout my afternoon.

drewnoakes commented:
> The biggest issue I had is the inability to use named arguments. Switching to tuples would allow this.

Ok, re-reading this I take it to mean that you're less fussed about quoting than being able to name arguments.

If that's the case, it's a style question. Let's look at some examples:

```c#
// CURRENT

var args = new ArgumentBuilder
{
    "blame",
    "--porcelain",
    { AppSettings.DetectCopyInFileOnBlame, "-M" },
    { AppSettings.DetectCopyInAllOnBlame, "-C" },
    { AppSettings.IgnoreWhitespaceOnBlame, "-w" },
    "-l",
    { lines != null, $"-L {lines}" },
    from.ToPosixPath().Quote(),
    "--",
    fileName.ToPosixPath().Quote()
};

// WITH TUPLES BUT NO ARG NAMES

var args = new ArgumentBuilder
{
    "blame",
    "--porcelain",
    (AppSettings.DetectCopyInFileOnBlame, "-M"),
    (AppSettings.DetectCopyInAllOnBlame, "-C"),
    (AppSettings.IgnoreWhitespaceOnBlame, "-w"),
    "-l",
    (lines != null, $"-L {lines}"),
    from.ToPosixPath().Quote(),
    "--",
    fileName.ToPosixPath().Quote()
};

// WITH TUPLES AND ARG NAMES

var args = new ArgumentBuilder
{
    "blame",
    "--porcelain",
    (condition: AppSettings.DetectCopyInFileOnBlame, ifTrue: "-M"),
    (condition: AppSettings.DetectCopyInAllOnBlame, ifTrue: "-C"),
    (condition: AppSettings.IgnoreWhitespaceOnBlame, ifTrue: "-w"),
    "-l",
    (condition: lines != null, ifTrue: $"-L {lines}"),
    from.ToPosixPath().Quote(),
    "--",
    fileName.ToPosixPath().Quote()
};
```

What do people think?

I find the difference between 1 and 2 minimal enough that we could go for it and use the names on a case-by-case basis.

Either way this is a simple refactor now or later.

sharwell commented:
> Ok, re-reading this I take it to mean that you're less fussed about quoting than being able to name arguments.

Both were bothering me enough to not be fully on board without a discussion. Maybe not a change, but at least a discussion. Fortunately, for as diverse a set of backgrounds as this project's maintainers have, everyone seems to do a remarkable job at seeing these topics through a reasonable multi-sided discussion to the end.

> I will look into it. I can't see how tuples would help as the names of tuple arguments don't seem to affect overload resolution.

Your interpretation is correct. But...

> What do people think?

I like the ability to include explicit names for literal arguments. For most of the examples you included, the only argument I would specify is `ifTrue`. For me, it's a big help to readability at the use site by distinguishing the following:

1. A pair of arguments that must be included in sequence, such as `-key value`
2. A key/value argument, such as `-key:value`
3. A single argument which is conditionally included

sharwell commented:
> I started looking at this and came up with a concern. I wonder about a form like this:
>
> ```
> git foo --bar="two words"
> ```

@drewnoakes For git, is it generally acceptable to pass `"--bar=two words"` for this case?

drewnoakes commented:
> For git, is it generally acceptable to pass "--bar=two words" for this case?

I'm not sure. I haven't found such an argument yet. I'm just being cautious.

Others here might be more familiar with the range of command line arguments.

drewnoakes commented:
I took an look at using tuples and I'm on the fence. The code becomes more dense and harder to read, for my eyes at least. I found adding `istrue` contributed to the density.

I'd be inclined to make that API change into another PR and discuss it separately.

There's some more valuable stuff blocked behind this change that I'd like to progress with.

sharwell commented:
@drewnoakes Thanks for looking into it. One more question: did you look to see if other projects are available that offer this functionality in a library?

drewnoakes commented:
> did you look to see if other projects are available that offer this functionality in a library?

I did have a look, but didn't find anything. There's lots of options for _parsing_. 

For something this simple where it's only really a few classes and the effort goes into getting a nice API then I would tend to bundle custom code and eliminate a dependency. If it turns out to be reusable, then it bulges out into a library.

Have you seen code for building arguments before? I'm not sure it'd be worth packaging these classes up as a library, as different applications will have different needs/idioms, especially if we get into value-add stuff like automatic quoting.

So long as it's project-internal we can evolve it, even with API breaking changes such as moving to tuples.

sharwell commented:
Thanks, did you see this question?

https://github.com/gitextensions/gitextensions/pull/4662#discussion_r178015051

drewnoakes commented:
> Thanks, did you see this question?

Oh cheers, no I missed it.

The problem was that the extension method's operations against the copied struct had no observable effect on the original. This is because I made initialisation of the `StringBuilder` field lazy (due to restrictions upon parameterless struct constructors), so if the first call was via an extension (quite likely) then any assignment to that field in the copy was not reflected in the original.

If we use a factory method, we cannot use the initialiser syntax, which would be a shame as it's quite nice.

If you know another route here I'd like to hear it. The only routes I can think of get us back to heap allocation, which nullifies the point of converting to a struct.

If the extension methods could take a `this ref ArgumentBuilder` parameter, it'd work as expected.

sharwell commented:
> If the extension methods could take a `this ref ArgumentBuilder` parameter, it'd work as expected.

They can, as long as you write it as `ref this` 😄

drewnoakes commented:
😆 I was surprised because I found [this](https://github.com/dotnet/csharplang/issues/186) which suggested it was in 7.2 but none of the checklist was ticked and it didn't seem to work. Obviously I didn't try hard enough.

[This issue](https://github.com/dotnet/roslyn/pull/23643) suggests the ordering constraint has been relaxed, but obviously not in my compiler.

Thanks for following up. I'll make the change and push a bit later tonight.

drewnoakes commented:
Just tried this again and glad to report I am not crazy. It doesn't work.

```c#
ref this ArgumentBuilder builder
```

...gives...

> CS1510: A ref or out value must be an assignable variable

Further digging:

```c#
// works
var args = new ArgumentBuilder();
args.Add(true, "hello");

// CS1510
var args = new ArgumentBuilder
{
    { true, "hello" }
};
```

What would happen if the extension method assigned to the "rvalue"? Not sure what Roslyn calls temporaries. In C++11 you can assign to rvalues. You don't have many opportunities to do that in C# though. This might be the first I've seen.

---

Thought this through and tried making `builder` an `in` parameter. It compiled and I was hopeful, but the tests still fail. Can't see how to get around this for now.

I'm more confused than when I started this little detour. I thought `in` was a kind of unassignable `ref` param, which feels like a perfect fit here. Why doesn't it work? If only I could ask one of the compiler peeps... 😉

drewnoakes commented:
Oh and the `ref this` thing was actually a limitation of R#. It highlights it as an error, even though it compiles. It's the caller that triggers the compile error in that case. I've [filed an issue](https://youtrack.jetbrains.com/issue/RSRP-469006).

sharwell commented:
👍 Thanks for the details. Can you file a bug with that repro example on dotnet/roslyn?

RussKie commented:
Great work
Anything we can squash?

drewnoakes commented:
> Can you file a bug?

https://github.com/dotnet/roslyn/issues/25862

> Anything we can squash?

Yep, will do so soon. Also enjoying a long weekend here.

drewnoakes commented:
> Anything we can squash?

Squashed from 16 down to 7 commits. Basically all new code ended up in one commit, as it's uninteresting to see how it came together. Other commits are discrete tweaks and adding unit tests to prior code.

RussKie commented:
great,  thank you

On Sun, 1 Apr 2018 7:05 am Drew Noakes <notifications@github.com> wrote:

> Anything we can squash?
>
> Squashed from 16 down to 7 commits. Basically all new code ended up in one
> commit, as it's uninteresting to see how it came together. Other commits
> are discrete tweaks and adding unit tests to prior code.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/gitextensions/gitextensions/pull/4662#issuecomment-377723379>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AEMyXsO5mg5Tar6DvAsj_RE0j93PWtosks5tj--MgaJpZM4SvN80>
> .
>

drewnoakes commented:
No worries. I'm turning in now but if I get a chance tomorrow I'd like to start putting together a PR that covers the git process stuff I've been looking at. Would be handy to merge this and rebase on top of it. Let me know if you or anyone else has any reservations. If not, I'll likely merge it some time tomorrow, UK time.
